scratchml/models/svr.py

```diff
@@ -0,0 +1,200 @@
+        self.X_ = X
+        self.y_ = y
+
+        # Set gamma if necessary
+        if self.gamma == 'scale':
+            self.gamma_ = 1 / (X.shape[1] * X.var())
+        elif self.gamma == 'auto':
+            self.gamma_ = 1 / X.shape[1]
+        else:
+            self.gamma_ = float(self.gamma)
+
+        # Compute the kernel matrix
+        self.K_ = self._kernel_function(X, X)
+
+        n_samples = X.shape[0]
+        K = self.K_
+
+        # Create P matrix
+        P_top = np.hstack((K, -K))
+        P_bottom = np.hstack((-K, K))
+        P = np.vstack((P_top, P_bottom))
+
+        # Ensure P is positive semi-definite
+        P = P.astype(np.float64) + 1e-8 * np.eye(2 * n_samples)
+
+        # Create q vector
+        q = np.hstack([self.epsilon - y, self.epsilon + y])
+
+        # Create G matrix and h vector for inequality constraints
+        G_std = np.vstack((-np.eye(2 * n_samples), np.eye(2 * n_samples)))
+        h_std = np.hstack((np.zeros(2 * n_samples), self.C * np.ones(2 * n_samples)))
+
+        # Create A matrix and b vector for equality constraint
+        A = np.hstack((np.ones(n_samples), -np.ones(n_samples))).reshape(1, -1)
+        b = np.array([0.0])
+
+        # Convert to cvxopt matrices
+        from cvxopt import matrix, solvers
+        P = matrix(P)
+        q = matrix(q)
+        G = matrix(G_std)
+        h = matrix(h_std)
+        A = matrix(A)
+        b = matrix(b)
+
+        # Solve QP problem
+        solvers.options['show_progress'] = False  # Suppress output
+        solution = solvers.qp(P, q, G, h, A, b)
+
+        # Extract alphas
+        z = np.array(solution['x']).flatten()
+        alpha = z[:n_samples]
+        alpha_star = z[n_samples:]
+        self.alphas_ = alpha - alpha_star
+
+        # Compute bias term
+        f = self.K_ @ self.alphas_
+        idx = np.where((alpha > 1e-5) & (alpha < self.C - 1e-5))[0]
+        idx_star = np.where((alpha_star > 1e-5) & (alpha_star < self.C - 1e-5))[0]
+
+        b_list = []
+
+        for i in idx:
+            b_i = y[i] - f[i] - self.epsilon
+            b_list.append(b_i)
+
+        for i in idx_star:
+            b_i = y[i] - f[i] + self.epsilon
+            b_list.append(b_i)
+
+        if b_list:
+            self.b_ = np.mean(b_list)
+        else:
+            self.b_ = 0.0  # Default to zero if no support vectors found
+
+    def predict(self, X):
+        """
+        Predicts target values for the given feature matrix.
+
+        Args:
+            X (np.ndarray): Feature matrix.
+
+        Returns:
+            np.ndarray: Predicted target values.
+        """
+        if self.X_ is None or self.alphas_ is None:
+            raise ValueError("The model has not been trained yet. Please call the fit method first.")
+
+        X = convert_array_numpy(X)
+        K_pred = self._kernel_function(self.X_, X)
+        predictions = (self.alphas_ @ K_pred) + self.b_
+        return predictions
+
+    def score(self, X, y, metric="r_squared"):
```

Comment:
Please, add typing here. Like this:

```python3
def score(self, X: np.ndarray, y: np.ndarray, metric: str = "r_squared") -> np.float64:
```

Reply:
You can also check out the `score` function in the Random Forest implementation to get an idea (line 547).